crates/networking/p2p/rlpx/connection.rs

```diff
@@ -506,168 +492,131 @@ impl<S: AsyncWrite + AsyncRead + std::marker::Unpin> RLPxConnection<S> {
-            Err(RLPxError::InvalidState())
-        }
+    async fn receive_auth(&mut self) -> Result<RemoteState, RLPxError> {
+        let secret_key: SecretKey = self.signer.clone().into();
+
+        let msg_bytes = self.receive_handshake_msg().await?;
+        let size_data = &msg_bytes
+            .get(..2)
+            .ok_or(RLPxError::InvalidMessageLength())?;
+        let msg = &msg_bytes
+            .get(2..)
+            .ok_or(RLPxError::InvalidMessageLength())?;
+        let (auth, remote_ephemeral_key) = decode_auth_message(&secret_key, msg, size_data)?;
+        self.remote_node_id = auth.node_id;
+
+        Ok(RemoteState {
+            nonce: auth.nonce,
+            ephemeral_key: remote_ephemeral_key,
+            init_message: msg_bytes.to_owned(),
+        })
     }
 
-    async fn receive_ack(&mut self) -> Result<(), RLPxError> {
-        if let RLPxConnectionState::InitiatedAuth(initiated_auth_state) = &self.state {
-            let secret_key: SecretKey = self.signer.clone().into();
-            // Clonning previous state to avoid ownership issues
-            let previous_state = initiated_auth_state.clone();
-            let msg_bytes = self.receive_handshake_msg().await?;
-            let size_data = &msg_bytes
-                .get(..2)
-                .ok_or(RLPxError::InvalidMessageLength())?;
-            let msg = &msg_bytes
-                .get(2..)
-                .ok_or(RLPxError::InvalidMessageLength())?;
-            let ack = decode_ack_message(&secret_key, msg, size_data)?;
-            let remote_ephemeral_key = ack
-                .get_ephemeral_pubkey()
-                .ok_or(RLPxError::NotFound("Remote ephemeral key".to_string()))?;
-            // Build next state
-            self.state = RLPxConnectionState::Established(Box::new(Established::for_initiator(
-                previous_state,
-                msg_bytes.to_owned(),
-                ack.nonce,
-                remote_ephemeral_key,
-            )));
-            Ok(())
-        } else {
-            Err(RLPxError::InvalidState())
-        }
+    async fn receive_ack(&mut self, local_state: LocalState) -> Result<(), RLPxError> {
+        let secret_key: SecretKey = self.signer.clone().into();
+        let msg_bytes = self.receive_handshake_msg().await?;
+        let size_data = &msg_bytes
+            .get(..2)
+            .ok_or(RLPxError::InvalidMessageLength())?;
+        let msg = &msg_bytes
+            .get(2..)
+            .ok_or(RLPxError::InvalidMessageLength())?;
+        let ack = decode_ack_message(&secret_key, msg, size_data)?;
+        let remote_ephemeral_key = ack
+            .get_ephemeral_pubkey()
+            .ok_or(RLPxError::NotFound("Remote ephemeral key".to_string()))?;
+
+        // keccak256(nonce || initiator-nonce)
+        // Local node is initator
+        let hashed_nonces = Keccak256::digest([ack.nonce.0, local_state.nonce.0].concat()).into();
+
+        self.framed.codec_mut().update_secrets(
```

Comment:
nit: maybe we can make `receive_ack` return remote state, create local state in the handshake and then just have update secrets receive both states instead of all this arguments? It also decouples the ack/auth messaging from the secrets which is more of a "handshake just finished" thing.

Reply:
I like this a lot!